Nullptr crash in WebCore::SecurityContext::policyContainer

pratiksha-99 opened the pull request:
#### d49fce1a2c4acc300a3b77a5c3e4d634a452ceff
<pre>
Nullptr crash in WebCore::SecurityContext::policyContainer
<a href="https://bugs.webkit.org/show_bug.cgi?id=281548">https://bugs.webkit.org/show_bug.cgi?id=281548</a>
<a href="https://rdar.apple.com/137177434">rdar://137177434</a>

Reviewed by Sihui Liu and Ryosuke Niwa.

Added a nullptr check for contentSecurityPolicy() before fetching its policyContainer().

* LayoutTests/fast/files/filereader-invalid-policy-container-expected.txt: Added.
* LayoutTests/fast/files/filereader-invalid-policy-container.html: Added.
* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::start):

Canonical link: <a href="https://commits.webkit.org/285316@main">https://commits.webkit.org/285316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e842b9342d685b886190494c72b3e6b945c6fdc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23252 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15436 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62191 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43445 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19665 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21780 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78066 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19186 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16509 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62215 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15953 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12879 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6521 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47440 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2224 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48509 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49797 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48252 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/fileapi/FileReaderLoader.cpp:
```diff
@@ -89,6 +89,11 @@ void FileReaderLoader::start(ScriptExecutionContext* scriptExecutionContext, con
         return;
     }
+
+    auto* contentSecurityPolicy = scriptExecutionContext->contentSecurityPolicy();
```
`CheckedPtr contentSecurityPolicy = scriptExecutionContext->contentSecurityPolicy();`

pratiksha-99 replied:
hmm.. it doesn't build for `CheckedPtr` type. I can change it to `ContentSecurityPolicy*` type to be consistent.

szewai replied:
You may need to add `#include "ContentSecurityPolicy.h"`. It's not about specifying the type, but using smart pointers if possible.

rniwa commented on LayoutTests/fast/files/filereader-invalid-policy-container.html:
```diff
@@ -0,0 +1,83 @@
+        })();
+    </script>
+</body>
+</html>
```
This test can be simplified to:
```
<!DOCTYPE html>
<html lang="en">
<body>
<p>This test passes if WebKit does not crash.</p>
<script>
if (window.testRunner) {
    testRunner.dumpAsText();
    testRunner.waitUntilDone();
}

(async () => {
    const canvasElement = HTMLDocument.parseHTMLUnsafe('').createElement('canvas');
    const offscreenCanvas = canvasElement.transferControlToOffscreen();
    await (await offscreenCanvas.getContext('2d').canvas.convertToBlob({})).arrayBuffer();

    globalThis.testRunner?.notifyDone();
})();
</script>
</body>
</html>
```

szewai commented on Source/WebCore/fileapi/FileReaderLoader.cpp:
```diff
@@ -89,6 +90,11 @@ void FileReaderLoader::start(ScriptExecutionContext* scriptExecutionContext, con
         return;
     }
+
+    ContentSecurityPolicy* contentSecurityPolicy = scriptExecutionContext->contentSecurityPolicy();
```
`CheckedPtr`

pratiksha-99 replied:
oops, missed it. fixed now.

pratiksha-99 committed:
Nullptr crash in WebCore::SecurityContext::policyContainer
https://bugs.webkit.org/show_bug.cgi?id=281548
rdar://137177434

Reviewed by Sihui Liu and Ryosuke Niwa.

Added a nullptr check for contentSecurityPolicy() before fetching its policyContainer().

* LayoutTests/fast/files/filereader-invalid-policy-container-expected.txt: Added.
* LayoutTests/fast/files/filereader-invalid-policy-container.html: Added.
* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::start):

Canonical link: https://commits.webkit.org/285316@main
